src/safeds_runner/server/memoization_map.py

```diff
@@ -0,0 +1,176 @@
+    computation_time
+        Duration the computation of the value took in nanoseconds
+    lookup_time
+        Duration the lookup of the value took in nanoseconds (key comparison + IPC)
+    memory_size
+        Amount of memory the memoized value takes up in bytes
+    """
+
+    last_access: int
+    computation_time: int
+    lookup_time: int
+    memory_size: int
+
+    def __str__(self) -> str:
+        """
+        Summarizes stats contained in this object.
+
+        Returns
+        -------
+        Summary of stats
+        """
+        return (  # pragma: no cover
+            f"Last access: {self.last_access}, computation time: {self.computation_time}, lookup time:"
+            f" {self.lookup_time}, memory size: {self.memory_size}"
+        )
+
+
+class MemoizationMap:
+    """
+    The memoization map handles memoized function calls.
+
+    This contains looking up stored values, computing new values if needed and calculating and updating statistics.
+    """
+
+    def __init__(
+        self,
+        map_values: dict[tuple[str, tuple[Any], tuple[Any]], Any],
+        map_stats: dict[tuple[str, tuple[Any], tuple[Any]], MemoizationStats],
+    ):
+        """
+        Create a new memoization map using a value store dictionary and a stats dictionary.
+
+        Parameters
+        ----------
+        map_values
+            Value store dictionary
+        map_stats
+            Stats dictionary
+        """
+        self.map_values: dict[tuple[str, tuple[Any], tuple[Any]], Any] = map_values
+        self.map_stats: dict[tuple[str, tuple[Any], tuple[Any]], MemoizationStats] = map_stats
+
+    def memoized_function_call(
+        self,
+        function_name: str,
+        function_callable: Callable,
+        parameters: list[Any],
+        hidden_parameters: list[Any],
+    ) -> Any:
+        """
+        Handle a memoized function call.
+
+        Looks up the stored value, determined by function name, parameters and hidden parameters and returns it if found.
+        If no value is found, computes the value using the provided callable and stores it in the map.
+        Every call to this function will update the memoization stats.
+
+        Parameters
+        ----------
+        function_name
+            Fully qualified function name
+        function_callable
+            Function that is called and memoized if the result was not found in the memoization map
+        parameters
+            List of parameters passed to the function
+        hidden_parameters
+            List of hidden parameters for the function. This is used for memoizing some impure functions.
+
+        Returns
+        -------
+        The result of the specified function, if any exists
+        """
+        key = (function_name, _convert_list_to_tuple(parameters), _convert_list_to_tuple(hidden_parameters))
+        time_compare_start = time.perf_counter_ns()
```

Comment:
I'd also consider the computation of the key as part of the [time_compare_start, time_compare_end] interval.